File: src/main/scala/search/algorithm/AhoCorasic.scala
package search.algorithm

import java.lang.Byte.toUnsignedInt

import search.engine.{MultiSearchContext, MultiSearchProcessor}

import scala.annotation.varargs
import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer


object AhoCorasic extends MultiSearchAlgorithm {

  final class Processor(
      private[this] val jumpTable: Array[Int],
      private[this] val matchFor: Array[Int],
      private[this] val needleLengths: Array[Int]) extends MultiSearchProcessor {

    private[this] var currentPosition = 0

    override def process(value: Byte): Boolean = {
      currentPosition = jumpTable(currentPosition | toUnsignedInt(value))
      if (currentPosition >= 0) true
      else {
        currentPosition = -currentPosition
        false
      }
    }

    override def reset(): Unit = {
      currentPosition = 0
    }

    override def needleLength: Int = {
      val foundNeedleId = getFoundNeedleId
      if (foundNeedleId >= 0) needleLengths(foundNeedleId) else 0
    }

    override def getFoundNeedleId: Int = matchFor(currentPosition >>> BitsPerSymbol)

  }

  final class Context(
      jumpTable: Array[Int],
      matchFor: Array[Int],
      needleLengths: Array[Int]) extends MultiSearchContext {

      override def newProcessor: Processor = new Processor(jumpTable, matchFor, needleLengths)

  }

  private final val BitsPerSymbol = 8
  private final val AlphabetSize = 1 << BitsPerSymbol

  @varargs
  override def apply(needles: Array[Byte]*): Context = {

    require(needles.forall(_.nonEmpty), "Cannot search for an empty needle.")

    val (jumpTable, matchFor) = buildJumpTable(needles)
    linkSuffixes(jumpTable, matchFor)

    for (i <- jumpTable.indices) {
      if (matchFor(jumpTable(i) >>> BitsPerSymbol) >= 0) {
        jumpTable(i) = -jumpTable(i)
      }
    }

    new Context(jumpTable, matchFor, needles.toArray.map(_.length))
  }

  override def apply(needle: Array[Byte]): Context = apply(Seq(needle): _*)

  private def buildJumpTable(needles: Seq[Array[Byte]]): (Array[Int], Array[Int]) = {

    val emptyJumpTableSegment = Array.fill(AlphabetSize)(-1)
    val jumpTableBuilder = ArrayBuffer(emptyJumpTableSegment: _*)
    val matchForBuilder = ArrayBuffer(-1)

    for (stringId <- needles.indices) {

      val str = needles(stringId)
      var currentPosition = 0

      for (signedByte <- str) {

        val next = currentPosition + toUnsignedInt(signedByte)

        if (jumpTableBuilder(next) == -1) {
          jumpTableBuilder(next) = matchForBuilder.size << BitsPerSymbol
          jumpTableBuilder ++= emptyJumpTableSegment
          matchForBuilder += -1
        }

        currentPosition = jumpTableBuilder(next)
      }

      matchForBuilder(currentPosition >>> BitsPerSymbol) = stringId
    }

    (jumpTableBuilder.toArray, matchForBuilder.toArray)
  }

  private def linkSuffixes(jumpTable: Array[Int], matchFor: Array[Int]): Unit = {

    val queue = mutable.Queue(0)
    val suffixLinks = Array.fill(matchFor.length)(-1)

    while (queue.nonEmpty) {

      val v = queue.dequeue()
      val u = if (v == 0) 0 else suffixLinks(v >>> BitsPerSymbol)

      if (matchFor(v >>> BitsPerSymbol) == -1) matchFor(v >>> BitsPerSymbol) = matchFor(u >>> BitsPerSymbol)

      for (ch <- 0 until AlphabetSize) {

        val vIndex = v | ch
        val uIndex = u | ch

        if (jumpTable(vIndex) != -1) {
          suffixLinks(jumpTable(vIndex) >>> BitsPerSymbol) = if (v > 0 && jumpTable(uIndex) != -1) jumpTable(uIndex) else 0
          queue += jumpTable(vIndex)
        } else {
          jumpTable(vIndex) = if (jumpTable(uIndex) != -1) jumpTable(uIndex) else 0
        }
      }
    }
  }

}

File: src/main/scala/search/algorithm/AhoCorasic.java
package search.algorithm;

import com.google.common.base.Preconditions;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Queue;

public final class AhoCorasic implements MultiSearchAlgorithm {

  private static final int BITS_PER_SYMBOL = 8;
  private static final int ALPHABET_SIZE = 1 << BITS_PER_SYMBOL;

  private final int[] jumpTable;
  private final int[] matchForNeedleId;
  private final int[] needleLengths;

  private final static class Processor implements MultiSearchProcessor {

    private final int[] jumpTable;
    private final int[] matchForNeedleId;
    private final int[] needleLengths;

    private int currentPosition;

    private Processor(int[] jumpTable, int[] matchForNeedleId, int[] needleLengths) {
      this.jumpTable = jumpTable;
      this.matchForNeedleId = matchForNeedleId;
      this.needleLengths = needleLengths;
    }

    @Override
    public boolean process(byte value) {
      currentPosition = jumpTable[currentPosition | Byte.toUnsignedInt(value)];
      if (currentPosition < 0) {
        currentPosition = -currentPosition;
        return false;
      }
      return true;
    }

    @Override
    public int getFoundNeedleId() {
      return matchForNeedleId[currentPosition >> BITS_PER_SYMBOL];
    }

    @Override
    public void reset() {
      currentPosition = 0;
    }

    @Override
    public int needleLength() {
      int foundNeedleId = getFoundNeedleId();
      return foundNeedleId >= 0 ? needleLengths[foundNeedleId] : 0;
    }

  }

  private AhoCorasic(byte[]... needles) {

    for (byte[] needle : needles) {
      Preconditions.checkArgument(needle.length > 0, "Needle must be non empty");
    }

    ArrayList<Integer> jumpTableBuilder = new ArrayList<>(ALPHABET_SIZE);
    for (int i = 0; i < ALPHABET_SIZE; i++) {
      jumpTableBuilder.add(-1);
    }

    ArrayList<Integer> matchForBuilder = new ArrayList<>();
    matchForBuilder.add(-1);

    needleLengths = new int[needles.length];

    for (int needleId = 0; needleId < needles.length; needleId++) {
      byte[] needle = needles[needleId];
      needleLengths[needleId] = needle.length;
      int currentPosition = 0;

      for (byte ch0 : needle) {

        final int ch = Byte.toUnsignedInt(ch0);
        final int next = currentPosition + ch;

        if (jumpTableBuilder.get(next) == -1) {
          jumpTableBuilder.set(next, jumpTableBuilder.size());
          for (int i = 0; i < ALPHABET_SIZE; i++) {
            jumpTableBuilder.add(-1);
          }
          matchForBuilder.add(-1);
        }

        currentPosition = jumpTableBuilder.get(next);
      }

      matchForBuilder.set(currentPosition >> BITS_PER_SYMBOL, needleId);
    }

    jumpTable = new int[jumpTableBuilder.size()];
    matchForNeedleId = new int[matchForBuilder.size()];

    for (int i = 0; i < jumpTableBuilder.size(); i++) {
      jumpTable[i] = jumpTableBuilder.get(i);
    }

    for (int i = 0; i < matchForBuilder.size(); i++) {
      matchForNeedleId[i] = matchForBuilder.get(i);
    }

    linkSuffixes();

    for (int i = 0; i < jumpTable.length; i++) {
      if (matchForNeedleId[jumpTable[i] >> BITS_PER_SYMBOL] >= 0) {
        jumpTable[i] = -jumpTable[i];
      }
    }
  }

  private void linkSuffixes() {

    Queue<Integer> queue = new ArrayDeque<>();
    queue.add(0);

    int[] suffixLinks = new int[matchForNeedleId.length];
    Arrays.fill(suffixLinks, -1);

    while (!queue.isEmpty()) {

      final int v = queue.remove();
      int vPosition = v >> BITS_PER_SYMBOL;
      final int u = suffixLinks[vPosition] == -1 ? 0 : suffixLinks[vPosition];

      if (matchForNeedleId[vPosition] == -1) {
        matchForNeedleId[vPosition] = matchForNeedleId[u >> BITS_PER_SYMBOL];
      }

      for (int ch = 0; ch < ALPHABET_SIZE; ch++) {

        final int vIndex = v | ch;
        final int uIndex = u | ch;

        final int jumpV = jumpTable[vIndex];
        final int jumpU = jumpTable[uIndex];

        if (jumpV != -1) {
          suffixLinks[jumpV >> BITS_PER_SYMBOL] = v > 0 && jumpU != -1 ? jumpU : 0;
          queue.add(jumpV);
        } else {
          jumpTable[vIndex] = jumpU != -1 ? jumpU : 0;
        }
      }
    }
  }

  @Override
  public MultiSearchProcessor newProcessor() {
    return new Processor(jumpTable, matchForNeedleId, needleLengths);
  }

  public static AhoCorasic init(byte[] ...needles) {
    return new AhoCorasic(needles);
  }

}
